packages/react-router/modules/matchPath.js

```diff
@@ -1,5 +1,59 @@
 import pathToRegexp from "path-to-regexp";
 
+function isAbsolute(pathname) {
+  return pathname.charAt(0) === "/";
+}
+
+function ensureTrailingSlash(pathname) {
+  return hasTrailingSlash(pathname) ? pathname : pathname + "/";
+}
+
+function hasTrailingSlash(pathname) {
+  return pathname.charAt(pathname.length - 1) === "/";
+}
+
+function isMalformed(pathname) {
+  return pathname.slice(0, 3) === "...";
+}
+
+function refersToParentSegment(pathname) {
+  return pathname.slice(0, 3) === "../";
+}
+
+function refersToCurrentSegment(pathname) {
+  return pathname.slice(0, 2) === "./";
+}
+
+function resolvePath(pathname, base) {
+  if (pathname === undefined || isAbsolute(pathname)) {
+    return pathname;
+  }
+
+  if (isMalformed(pathname)) {
+    throw new Error("cannot resolve pathname: pathname is malformed");
+  }
+
+  if (refersToParentSegment(pathname)) {
+    throw new Error(
+      "cannot resolve pathname: pathname refers to parent path-segment"
+    );
+  }
+
+  if (refersToCurrentSegment(pathname)) {
+    pathname = pathname.substr(2);
+  }
+
+  if (!base) {
+    base = "/";
+  }
+
+  if (pathname === "") {
+    return base;
+  } else {
+    return `${ensureTrailingSlash(base)}${pathname}`;
```

Comment:
Using string interpolation seems a bit convoluted for just a concatenation of variable strings:

`return ensureTrailingSlash(base) +  pathname`

Reply:
This is a part I have not touched. It is based on the commits by @lukaswelinder. Anyway, I think your solution is more clear.